### PR TITLE
Convert latte

### PR DESCRIPTION
**Pull request by Hehk**
I was doing this while trying to debug the `call of undefined` issue which looks to be coming from our implementation of latte.

- [x] Fix serving ts files onto remote
- [x] Finish the conversion

### CODE REVIEW

File: lib/latte.ts [hunk 1]
```diff
@@ -0,0 +1,482 @@
+// Latte supports Mocha syntax for now, but we'd like to evolve.
```
Comment (patrickdbakke) [inline lib/latte.ts:1]: @Hehk i'm curious if this conversion helped us learn more about zen... 
Did this conversion shed any light on why zen needs a separate version of latte, and can't just use `mocha` directly?

File: lib/latte.ts [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+      depth: previous.depth + 1,
+      testCount: 0,
+      fullName: [previous.fullName, name].filter((s) => !!s).join(' '),
+      before: [],
+      beforeEach: [],
+      after: [],
+      afterEach: [],
+      helpers: {},
+      // eslint-disable-next-line @typescript-eslint/no-empty-function
+      timeout: () => {}, // TODO
```
Comment (patrickdbakke) [inline lib/latte.ts:244]: Acknowledging that this was here before the TS conversion - what's this TODO about?

File: lib/latte.ts [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+    current = previous
+  }
+
+  window.describe = describe
+  window.context = window.describe
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  describe.skip = function () {}
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  describe.only = function () {} // TODO
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
```
Comment (patrickdbakke) [inline lib/latte.ts:263]: Nit: can use `function () {/**/}` in place of the eslint disable.

File: lib/latte.ts [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+    setTimeoutPromise(10000)
+
+    context.zen = {}
+    context.zen.extendRemoteTimeout = (ms: number) => {
+      clearTimeout(currentTimeout)
+      setTimeoutPromise(ms)
+    }
+
+    let runPromise = null
+    // TODO this is probably the point where 'call' is coming from
```
Comment (patrickdbakke) [inline lib/latte.ts:417]: This makes sense to me +++

File: lib/latte.ts [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+    if (!cbOrTest.fn) throw new Error('No valid test')
+
+    if (cbOrTest.fn.length > 0) {
+      runPromise = new RealPromise((res) => cbOrTest.fn?.call(context, res))
+    } else {
+      runPromise = cbOrTest.fn.call(context)
+    }
+
+    const race = RealPromise.race([timeoutPromise, runPromise])
+    return promiseFinally.call(race, () => {
```
Comment (patrickdbakke) [inline lib/latte.ts:427]: Nit: if we only care about chrome, we can drop the polyfill, and just `race.finally`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally

File: lib/latte.ts [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+        step = queue.pop()
+      }
+      return allTests
+    },
+    run: async function latteRun(tests) {
+      // Store logs so they can be relayed back from remote
+      const log: string[] = []
+      const originalConsoleLog = console.log
+      console.log = (...args) => {
+        log.push(args.join(' '))
```
Comment (patrickdbakke) [inline lib/latte.ts:146]: No changes needed... 
Can we make this work for non-string logs?

File: lib/latte.ts (hunk 1 above)
Comment (patrickdbakke) [reply lib/latte.ts:1]: After reviewing, nothing in here stands out to me as "obviously we can't use mocha because ____"

**Review by patrickdbakke — APPROVED** [review]

File: lib/latte.ts (hunk 2 above)
Comment (Hehk) [reply lib/latte.ts:244]: I believe this relates to the timeout fn we have defined in some of the tests itself. Idk the purpose, my best guess was maybe something like the extendRemoteTimeout I implemented a couple weeks ago 🤷

File: lib/latte.ts (hunk 6 above)
Comment (Hehk) [reply lib/latte.ts:146]: The next step outside of this file is to JSON.stringify this content. It is probably not worth it right now to make sure that all looks good, I'll make it better if there is an issue where that will help with

**Pull request merged**